File: Codes/2389-maximum-number-of-robots-within-budget.py
'''
Author: LetMeFly
Date: 2024-09-13 13:09:24
LastEditors: LetMeFly
LastEditTime: 2024-09-13 13:12:48
'''
from typing import List
from collections import deque

class Solution:
    def maximumRobots(self, chargeTimes: List[int], runningCosts: List[int], budget: int) -> int:
        ans = sumCosts = 0
        l = 0
        q = deque()
        for r in range(len(chargeTimes)):
            while q and chargeTimes[r] >= chargeTimes[q[-1]]:
                q.pop()
            q.append(r)
            sumCosts += runningCosts[r]

            while q and chargeTimes[q[0]] + (r - l + 1) * sumCosts > budget:
                if l == q[0]:
                    q.popleft()
                sumCosts -= runningCosts[l]
                l += 1
            
            ans = max(ans, r - l + 1)
        return ans

File: Codes/2389-maximum-number-of-robots-within-budget.go
package main

func maximumRobots(chargeTimes []int, runningCosts []int, budget int64) int {
    ans := 0
    sumCosts := int64(0)
    q := []int{}

    for l, r := 0, 0; r < len(chargeTimes); r++ {
        for len(q) > 0 && chargeTimes[r] >= chargeTimes[q[len(q) - 1]] {
            q = q[:len(q) - 1]
        }
        q = append(q, r)
        sumCosts += int64(runningCosts[r])
        
        for len(q) > 0 && int64(chargeTimes[q[0]]) + int64(r - l + 1) * sumCosts > budget {
            if l == q[0] {
                q = q[1:]
            }
            sumCosts -= int64(runningCosts[l])
            l++
        }

        if r - l + 1 > ans {
            ans = r - l + 1
        }
    }
    return ans
}